### PR TITLE
Caps Lock & Num Lock Tooltip Indicator

### PR DESCRIPTION
**Pull request by andirsun**
close : #84 
## Demo : 
deepin-screen-recorder_Select area_20201202221117

### CODE REVIEW

**Comment by andirsun** [comment]
I just improve the message as the same behavior in greeter. 
Case 1 : "Num Lock is on"
Case 2 : "Caps Lock is on"
Case 3 : "Caps Lock &amp; Num Lock are on"
deepin-screen-recorder_Select area_20201205113109

File: src/Indicator.vala [hunk 1]
```diff
@@ -171,6 +173,21 @@ public class Keyboard.Indicator : Wingpanel.Indicator {
             warning ("Error launching keyboard layout display: %s", e.message);
         }
     }
+
+    private void update_tooltip () {
+        string message = "";
+
+        if (keymap.get_num_lock_state () && settings.get_boolean ("numlock") &&
+            keymap.get_caps_lock_state () && settings.get_boolean ("capslock") ) {
+            message = "Caps Lock &amp; Num Lock are on";
```
Comment (meisenzahl) [inline src/Indicator.vala:182]:
```suggestion
            message = _("Caps Lock &amp; Num Lock are on");
```

File: src/Indicator.vala [hunk 2]
```diff
@@ -171,6 +173,21 @@ public class Keyboard.Indicator : Wingpanel.Indicator {
...
+        } else if (keymap.get_num_lock_state () && settings.get_boolean ("numlock")) {
+            message = "Num Lock is on";
```
Comment (meisenzahl) [inline src/Indicator.vala:184]:
```suggestion
            message = _("Num Lock is on");
```

File: src/Indicator.vala [hunk 3]
```diff
@@ -171,6 +173,21 @@ public class Keyboard.Indicator : Wingpanel.Indicator {
...
+        } else if (keymap.get_caps_lock_state () && settings.get_boolean ("capslock")) {
+            message = "Caps Lock is on";
```
Comment (meisenzahl) [inline src/Indicator.vala:186]:
```suggestion
            message = _("Caps Lock is on");
```

**Review by meisenzahl — CHANGES_REQUESTED** [review]
I suggested some changes to make the texts translatable.

**Review by meisenzahl — COMMENTED** [review]
LGTM but lets wait for review of @elementary/ux

**Review by cassidyjames — CHANGES_REQUESTED** [review]
Can we have different tooltips per icon? It would make it more clear what each indicator is showing. Then we would just have “Num Lock is on” for the numlock icon, and “Caps Lock is on” for the Caps Lock icon. No need to say they're off when the icon is not visible.

For the layout icon, the current layout name would make sense to show in a tooltip, but that can be a different PR.

**Comment by andirsun** [comment]
> Can we have different tooltips per icon? It would make it more clear what each indicator is showing. Then we would just have “Num Lock is on” for the numlock icon, and “Caps Lock is on” for the Caps Lock icon. No need to say they're off when the icon is not visible.
> 
> For the layout icon, the current layout name would make sense to show in a tooltip, but that can be a different PR.

Sure!, I can work on this change tomorrwo

**Comment by jeremypw** [comment]
On my system the caps lock and numlock revealers seem to operate weirdly.  Sometimes nothing happens - sometimes the wrong revealer activates. Investigating.

**Comment by jeremypw** [comment]
Sorry - just realised this is only about tooltips!

**Review by jeremypw — CHANGES_REQUESTED** [review]
Do you need to update the tooltips on visibility change?  When the child is not visible the tooltip does not change and is not visible.  I think they can just be set once as tooltips of the *icons* in the constructor and lose the update_tooltips () functions.

**Comment by andirsun** [comment]
> Do you need to update the tooltips on visibility change?  When the child is not visible the tooltip does not change and is not visible.  I think they can just be set once as tooltips of the *icons* in the constructor and lose the update_tooltips () functions.

You are right, the tooltips can be setted only in the constructor. Let me do the change.
Thanks for review

**Comment by jeremypw** [comment]
I think my other problems with these revealers may be due to running in a VirtualBox VM which seems to swallow some keymap state changes.

**Review by jeremypw — APPROVED** [review]
Looks good to me.

**Review by cassidyjames — APPROVED** [review]
Looks good UX-wise as well. Thanks!

**Pull request merged**